pkg/dcgm/field_values.go

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dcgm
+
+/*
+#include "dcgm_agent.h"
+#include "dcgm_structs.h"
+#include "field_values_cb.h"
+extern int go_dcgmFieldValueEntityEnumeration(dcgm_field_entity_group_t entityGroupId,
+            dcgm_field_eid_t entityId,
+            dcgmFieldValue_v1 *values,
+            int numValues,
+            void *userData);
+*/
+import "C"
+import (
+	"fmt"
+	"sync"
+	"time"
+	"unsafe"
+)
+
+// FieldValueEntity represents a field value entity
+type FieldValueEntity struct {
+	ts    int64
+	value [4096]byte
+}
+
+// AsInt64 returns field value as int64
+func (v FieldValueEntity) AsInt64() int64 {
+	ptr := (*C.int64_t)(unsafe.Pointer(&v.value[0]))
+	return int64(*ptr)
+}
+
+// FieldValueRecords represents field value records output in the format: [entityGroupId][entityId][fieldId][]FieldValue
+type FieldValues map[FieldEntityGroup]map[uint]map[int][]FieldValueEntity
+
+type callback struct {
+	mu     sync.Mutex
+	Values FieldValues
+}
+
+func (cb *callback) processValues(entityGroupID FieldEntityGroup, entityID uint, values []C.dcgmFieldValue_v1) {
+	if cb.Values == nil {
+		cb.Values = FieldValues{}
+	}
+
+	if _, exists := cb.Values[entityGroupID]; !exists {
+		cb.Values[entityGroupID] = map[uint]map[int][]FieldValueEntity{}
+	}
+
+	if _, exists := cb.Values[entityGroupID][entityID]; !exists {
+		cb.Values[entityGroupID][entityID] = map[int][]FieldValueEntity{}
+	}
+
+	cb.mu.Lock()
+
+	for _, v := range values {
+		if v.status > -1 {
```

Comment:
We have C.DCGM_ST_OK constant that should be used here. And it should be `==`, we do not expect seeing any positive value here.